### PR TITLE
Delete orphan blocks to avoid memory leak

### LINKED ISSUE #261: Orphan blocks are leaked on flush
The blocks which are not yet a part of the chain are considered orphan and stored in `mapOrphanBlocks` which is periodically flushed during sync. The problem is that it's storing naked pointers which aren't deleted on flush which should cause a memory leak (haven't checked in Valgrind yet).

Change this to store a shared pointer until the block has been serialized to disk and detached from the orphan map.

### PR DESCRIPTION
**Pull request by pwnorbitals**
First fix as a new developper !
In the future, make a switch to smart pointers in mapOrphanBlocks variable.

### CODE REVIEW

**Review by tomasbrod — APPROVED** [review]
Good work.
* the commit message could be better
* have you checked all occurrences of `mapOrphanBlocks.clear` and `.erase` and `[]=` ?

**Comment by pwnorbitals** [comment]
I had comments on slack about the commit message. I will edit it if I can, or otherwise won't do the error again
I checked all occurences of `.clear()`, but not `.erase` and `[]=`, you are right it may also leak memory, probably in smaller amounts though
Sorry for the beginner errors

File: src/main.cpp [hunk 1]
```diff
@@ -4678,14 +4678,19 @@ bool AskForOutstandingBlocks(uint256 hashStart)
 }
 
 
-
+void clearOrphanBlocks() {
```
Comment (denravonska) [inline src/main.cpp:4681]: Move the brackets down for consistency.

File: src/main.cpp [hunk 2]
```diff
@@ -4678,14 +4678,19 @@ bool AskForOutstandingBlocks(uint256 hashStart)
...
+    for(auto it = mapOrphanBlocks.begin(); it != mapOrphanBlocks.end(); it++) {
+        delete it->second;
+    }
+    mapOrphanBlocks.clear();
```
Comment (denravonska) [inline src/main.cpp:4685]: I think we have to clear `mapOrphanBlocksByPrev` here as well as it seems to hold the same pointer.

Comment (denravonska) [reply src/main.cpp:4685]: While we're at it, do a `LOCK(cs_main)` when entering the function.

**Comment by pwnorbitals** [comment]
New PR : https://github.com/gridcoin/Gridcoin-Research/pull/609

**Comment by tomasbrod** [comment]
Please, don't do this. It makes things hard to track. Next time (not now, plz), you can push new changes to your PR branch (you keep it in branch, right?), you can even force push to PRs if you have not closed them.

**Comment by pwnorbitals** [comment]
yes, but I had deleted the PR branch by mistake. I think I had no other choice :( 
look : `MagixInTheAir  wants to merge 2 commits into gridcoin:development from unknown repository`

**Pull request closed without merging**